### DISCUSSION TITLE
rtsp push from IP-Camera: invalid track path

### DISCUSSION THREAD

**Discussion by tom1235465** [bluenviron/mediamtx#822] [Questions] [answered]
## Which version are you using?

v0.17.16

## Which operating system are you using?

- [x] Linux arm7 standard

## Describe the issue

The rtsp push by an IP-Camera to the rtsp-simple-server leads to "[RTSP] [conn 192.168.1.174:2062] closed (invalid track path (live/trackID=1))"

I want to connect ip-cameras to the rtsp-simple-server by using the rtsp push service.
As the camera doesn't offer rtsp push officially but uses it to connect to their own server (running Wowza Media Server 3), it seems to me like they didn't care much about the right sdp format. "s=","i=", "a=x-qt-text..." are all empty and I can't change that.
To initiate the push I'm only able to change: ip, port, path, username, password. Username and password can't be empty. The problem consists with and without usersettings in the server settings.
The camera also has an rtsp stream on port 554 running. It works when I put the link into the server-config. To avoid port forwarding and offering the stream to everyone, I would much rather prefer the push-option.

I also tested it with the (rtsp-) wowza steaming cloud, that works.

Does anyone have an idea what causes the "invalid track path"?

Some info about the camera stream:
Video:
H264
mimeType='video/H264' 
sampleRate='15'
width='640' 
height='480'
SPS=J0KAHtoCgPTAVSAAAAMAIAAAB4wIAD0IAA9CV73wvCIRqA==
PPS=KM48gA==

Audio: 
AAC
mimeType='audio/aac' 
sampleRate='8e+03'>
sampleCount='1024' 
channelCount='1'>
DataRate target='16000'

I try to 
1. start the server
2. publish with rtsp push
3. read with hls

## Did you attach the server logs?

yes: 
192.168.1.174 is the camera, 192.168.1.58 is a raspberry pi with rtsp-simple-server. I also tried it on a windows machine. 

2022/02/10 12:54:20 INF rtsp-simple-server v0.17.16
2022/02/10 12:54:20 DEB path manager opened
2022/02/10 12:54:20 INF [RTSP] listener opened on :8554 (TCP)
2022/02/10 12:54:20 INF [HLS] listener opened on :8888
2022/02/10 12:54:31 INF [RTSP] [conn 192.168.1.174:3840] opened
2022/02/10 12:54:31 DEB [RTSP] [conn 192.168.1.174:3840] [c->s] ANNOUNCE rtsp://                                                                      192.168.1.58:8554/live RTSP/1.0
CSeq: 1
Content-Length: 590
Content-Type: application/sdp

v=0
o=- 249831380 249831380 IN IP4 127.0.0.1
s=
i=
c=IN IP4 192.168.1.58
t=0 0
a=x-qt-text-nam:
a=x-qt-text-inf:
a=x-qt-text-cmt:source application:Whee
a=x-qt-text-aut:
a=x-qt-text-cpy:
m=video 0 RTP/AVP 96
a=rtpmap:96 H264/90000
a=fmtp:96 packetization-mode=1;profile-level-id=42801E;sprop-parameter-sets=J0KA                                                                      HtoCgPTAVSAAAAMAIAAAB4wIAD0IAA9CV73wvCIRqA==,KM48gA==
a=control:trackID=1
m=audio 0 RTP/AVP 97
a=rtpmap:97 MPEG4-GENERIC/8000
a=fmtp:97 streamtype=5;profile-level-id=1;mode=AAC-hbr;sizelength=13;indexlength                                                                      =3;indexdeltalength=3;config=1588
a=control:trackID=2

2022/02/10 12:54:31 INF [RTSP] [session 214391818] opened by 192.168.1.174:3840
2022/02/10 12:54:31 DEB [RTSP] [session 214391818] non-critical authentication e                                                                      rror:
2022/02/10 12:54:31 DEB [RTSP] [conn 192.168.1.174:3840] [s->c] RTSP/1.0 401 Una                                                                      uthorized
CSeq: 1
Server: gortsplib
WWW-Authenticate: Basic realm="IPCAM"
WWW-Authenticate: Digest realm="IPCAM", nonce="3e082d6046a026ec7e95d5564f0594d1"

2022/02/10 12:54:31 DEB [RTSP] [conn 192.168.1.174:3840] [c->s] ANNOUNCE rtsp://                                                                      192.168.1.58:8554/live RTSP/1.0
Authorization: Basic dGVzdDp0ZXN0
CSeq: 2
Content-Length: 590
Content-Type: application/sdp

v=0
o=- 249831380 249831380 IN IP4 127.0.0.1
s=
i=
c=IN IP4 192.168.1.58
t=0 0
a=x-qt-text-nam:
a=x-qt-text-inf:
a=x-qt-text-cmt:source application:Whee
a=x-qt-text-aut:
a=x-qt-text-cpy:
m=video 0 RTP/AVP 96
a=rtpmap:96 H264/90000
a=fmtp:96 packetization-mode=1;profile-level-id=42801E;sprop-parameter-sets=J0KA                                                                      HtoCgPTAVSAAAAMAIAAAB4wIAD0IAA9CV73wvCIRqA==,KM48gA==
a=control:trackID=1
m=audio 0 RTP/AVP 97
a=rtpmap:97 MPEG4-GENERIC/8000
a=fmtp:97 streamtype=5;profile-level-id=1;mode=AAC-hbr;sizelength=13;indexlength                                                                      =3;indexdeltalength=3;config=1588
a=control:trackID=2

2022/02/10 12:54:31 INF [RTSP] [session 929896411] opened by 192.168.1.174:3840
2022/02/10 12:54:31 DEB [path live] opened
2022/02/10 12:54:31 DEB [RTSP] [conn 192.168.1.174:3840] [s->c] RTSP/1.0 200 OK
CSeq: 2
Server: gortsplib
Session: 883768684;timeout=60

2022/02/10 12:54:31 DEB [RTSP] [conn 192.168.1.174:3840] [c->s] SETUP rtsp://192                                                                      .168.1.58:8554/live/trackID=1 RTSP/1.0
Authorization: Basic dGVzdDp0ZXN0
CSeq: 3
Transport: RTP/AVP/TCP;unicast;mode=receive;interleaved=0-1
User-Agent: Whee (LIVE555 Streaming Media v2010.04.09)

2022/02/10 12:54:31 INF [RTSP] [session 103214943] opened by 192.168.1.174:3840
2022/02/10 12:54:31 DEB [RTSP] [conn 192.168.1.174:3840] [s->c] RTSP/1.0 400 Bad                                                                       Request
CSeq: 3
Server: gortsplib

2022/02/10 12:54:31 INF [RTSP] [conn 192.168.1.174:3840] closed (invalid track p                                                                      ath (live/trackID=1))
2022/02/10 12:54:31 INF [RTSP] [session 214391818] closed (not in use)
2022/02/10 12:54:31 DEB [path live] closed (not in use)
2022/02/10 12:54:31 INF [RTSP] [session 929896411] closed (not in use)
2022/02/10 12:54:31 INF [RTSP] [session 103214943] closed (not in use)

## Did you attach a network dump?

yes. If it helps, I can also create a network dump when connecting successfully to the woza cloud
[wireshark.zip](https://github.com/aler9/rtsp-simple-server/files/8040876/wireshark.zip)

**Comment by tom1235465** [comment]
I figured out that in the serversession.go this part gets skipped 
```
	for trackID, track := range announcedTracks {
		u, _ := track.track.url(setuppedBaseURL)
		if u.String() == url.String() {
			return trackID, *setuppedPath, *setuppedQuery, nil
		}
	}
```

If I push the camera stream directly to rtsp-simple-server, announcedTracks is empty: `[]`
If I use ffmpeg to read the camera rtsp stream and push it to rtsp-simple-server, it works and the  announcedTracks is `[{0xc000056720 <nil>} {0xc000056840 <nil>}]`

As I'm no coder, it would be very helpfull if anyone could tell me where announcedTracks gets its values.

Edit: 
Seems like we found it: at SETUP, the client doesn't send the Session#. 

Edit2:
Is that the reason why it starts the function "newServerSession" after SETUP again? And is it okay that "s" gets more "map" entries?
```
after ANNOUNCE: newServerSession: s= &{0xc0002e6000 10s 10s :8554 <nil>    0 0 512 2048 0xbee520 0xbee5c0 10000000000 60000000000 1000000000 0xc0002c7180 0xb81240 {{} [1 4 0]} <nil> [] 0xc000004630 <nil> <nil  map[2035279264:0xc000314600] map[0xc000397290:{}] <nil> map[] 0xc00042a780 0xc00042a7e0 0xc00042a840 0xc00042a8a0 0xc00042a900 0xc00042a960}
after    SETUP: newServerSession: s= &{0xc0002e6000 10s 10s :8554 <nil>    0 0 512 2048 0xbee520 0xbee5c0 10000000000 60000000000 1000000000 0xc0002c7180 0xb81240 {{} [1 5 0]} <nil> [] 0xc000004630 <nil> <nil> map[2035279264:0xc000314600 3412083719:0xc00017cb00] map[0xc000397290:{}] <nil> map[] 0xc00042a780 0xc00042a7e0 0xc00042a840 0xc00042a8a0 0xc00042a900 0xc00042a960}
```

At the next step, at PLAY, the client would send the Session# (seen at the connection to wowza).
Looking at RFC2326, it seems like a regular behavior to get (and use) the session# after the SETUP part:

> The server generates session identifiers in response to SETUP
>    requests.

https://datatracker.ietf.org/doc/html/rfc2326#page-33

Any ideas for a workaround? 
 
Thanks,
Tom

**Comment by tom1235465** [comment, marked as answer]
My brother was able to solve the problem.
I guess it could be a server bug as it expects the SessionID already at "SETUP", regarding RFC2326 it should be generated after "SETUP".
Our workaround: gortsplib got edited, now the client can set a sessionID by putting it into the path. Path /live/session/4378590432/test sets the sessionID to 4378590432. If there's no /session/ in the path, it uses the regular way to get a sessionID. 
If someone else would need it, we can share it, but it's quite a special case I guess.

The next problem is, the camera sends a "PLAY" instead of RECORD" and we get a "must be in state [prePlay play], while is in state preRecord". A quickfix was to change the cases Play and Record in serversession.go. 

So with some modifications we got it running, thank you very much for that awesome server! I got you some coffee ;)

**Comment by aler9** [comment]
Issue solved in #840

[a comment by an automated account is not shown]
